File: contract-agent/src/pairingClient.ts
import {
  AgentPairRequest,
  AgentPairResponse,
  AgentPairVerifyRequest,
  AgentPairVerifyResponse,
  AgentSessionRefreshRequest,
  AgentSessionRefreshResponse,
  TaskType,
  DeviceOs,
} from "@slash-agent/contracts";

export interface PairAgentParams {
  apiBaseUrl: string;
  pairingCode: string;
  publicKeyBase64: string;
  deviceName: string;
  supportedTaskTypes: TaskType[];
}

/** 페어링 REST 응답은 {data,meta} 봉투를 쓴다 (메시지 프로토콜 문서 §3.3). */
async function postJson<T>(url: string, body: unknown, headers?: Record<string, string>): Promise<T> {
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...headers },
    body: JSON.stringify(body),
  });
  const parsed = (await res.json().catch(() => null)) as { data: T } | { error: { code: string; message: string } } | null;
  if (!res.ok) {
    const message = parsed && "error" in parsed ? `${parsed.error.code}: ${parsed.error.message}` : `HTTP ${res.status}`;
    throw new Error(`POST ${url} failed: ${message}`);
  }
  return (parsed as { data: T }).data;
}

export async function pairAgent(params: PairAgentParams): Promise<AgentPairResponse> {
  const body: AgentPairRequest = {
    pairingCode: params.pairingCode,
    publicKey: params.publicKeyBase64,
    device: {
      name: params.deviceName,
      os: "MACOS" as DeviceOs,
      architecture: process.arch === "arm64" ? "ARM64" : "X86_64",
      osVersion: process.platform,
      agentVersion: "contract-agent-simulator/0.1.0",
    },
    supportedTaskTypes: params.supportedTaskTypes,
  };
  return postJson<AgentPairResponse>(`${params.apiBaseUrl}/api/v1/agent/pair`, body);
}

export async function verifyPairing(
  apiBaseUrl: string,
  body: AgentPairVerifyRequest
): Promise<AgentPairVerifyResponse> {
  return postJson<AgentPairVerifyResponse>(`${apiBaseUrl}/api/v1/agent/pair/verify`, body);
}

/** 재페어링 없이 기기 인증 토큰만 갱신한다 (메시지 프로토콜 문서 §8.1 3단계). */
export async function refreshSession(
  apiBaseUrl: string,
  currentDeviceToken: string,
  body: AgentSessionRefreshRequest
): Promise<AgentSessionRefreshResponse> {
  return postJson<AgentSessionRefreshResponse>(`${apiBaseUrl}/api/v1/agent/sessions/refresh`, body, {
    Authorization: `Bearer ${currentDeviceToken}`,
  });
}

File: slash-python-pc-runner/src/slash_pc_runner/pairing_client.py
"""HTTP 페어링/토큰 갱신 — pairingClient.ts 대응."""

from __future__ import annotations

import json
import platform
import urllib.error
import urllib.request
from typing import Optional

from ._build_info import get_agent_version


def _post_json(url: str, body: dict, headers: Optional[dict] = None) -> dict:
    """페어링 REST 응답은 {data,meta} 봉투를 쓴다 (메시지 프로토콜 문서 §3.3)."""
    data = json.dumps(body).encode("utf-8")
    request_headers = {"Content-Type": "application/json", **(headers or {})}
    req = urllib.request.Request(url, data=data, headers=request_headers, method="POST")
    try:
        with urllib.request.urlopen(req) as res:
            parsed = json.loads(res.read().decode("utf-8"))
            return parsed["data"]
    except urllib.error.HTTPError as e:
        body_text = e.read().decode("utf-8")
        try:
            parsed = json.loads(body_text)
            message = f"{parsed['error']['code']}: {parsed['error']['message']}"
        except (json.JSONDecodeError, KeyError):
            message = f"HTTP {e.code}"
        raise RuntimeError(f"POST {url} failed: {message}") from e


def _architecture() -> str:
    return "ARM64" if platform.machine() in ("arm64", "aarch64") else "X86_64"


def pair_agent(
    api_base_url: str,
    pairing_code: str,
    public_key_base64: str,
    device_name: str,
    supported_task_types: list[str],
) -> dict:
    body = {
        "pairingCode": pairing_code,
        "publicKey": public_key_base64,
        "device": {
            "name": device_name,
            "os": "MACOS",
            "architecture": _architecture(),
            "osVersion": platform.platform(),
            "agentVersion": get_agent_version(),
        },
        "supportedTaskTypes": supported_task_types,
    }
    return _post_json(f"{api_base_url}/api/v1/agent/pair", body)


def verify_pairing(api_base_url: str, pairing_session_id: str, challenge_id: str, signature: str) -> dict:
    body = {"pairingSessionId": pairing_session_id, "challengeId": challenge_id, "signature": signature}
    return _post_json(f"{api_base_url}/api/v1/agent/pair/verify", body)


def refresh_session(
    api_base_url: str,
    current_device_token: str,
    device_id: str,
    refresh_nonce: str,
    requested_at: str,
    signature: str,
) -> dict:
    """재페어링 없이 기기 인증 토큰만 갱신 (메시지 프로토콜 문서 §8.1 3단계)."""
    body = {"deviceId": device_id, "refreshNonce": refresh_nonce, "requestedAt": requested_at, "signature": signature}
    return _post_json(
        f"{api_base_url}/api/v1/agent/sessions/refresh",
        body,
        headers={"Authorization": f"Bearer {current_device_token}"},
    )
